File: internal/repository/invoice_repo.go
package repository

import (
	"database/sql"
	"fmt"

	"Rocket-Invo/internal/database"
	"Rocket-Invo/internal/models"
)

type InvoiceRepo struct {
	db *database.DB
}

func NewInvoiceRepo(db *database.DB) *InvoiceRepo {
	return &InvoiceRepo{db}
}

func (r *InvoiceRepo) Create(inv *models.Invoice) (int64, error) {
	tx, err := r.db.Begin()
	if err != nil {
		return 0, err
	}
	defer tx.Rollback()

	res, err := tx.Exec(`
		INSERT INTO invoices (invoice_no, customer_id, date, subtotal, tax_rate, tax_amount, discount, total, notes, status)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`, inv.InvoiceNo, inv.CustomerID, inv.Date, inv.Subtotal, inv.TaxRate, inv.TaxAmount, inv.Discount, inv.Total, inv.Notes, inv.Status)
	if err != nil {
		return 0, err
	}

	invoiceID, err := res.LastInsertId()
	if err != nil {
		return 0, err
	}

	for _, item := range inv.Items {
		_, err := tx.Exec(`
			INSERT INTO invoice_items (invoice_id, product_id, product_name, quantity, unit_price, total, rental_period)
			VALUES (?, ?, ?, ?, ?, ?, ?)
		`, invoiceID, item.ProductID, item.ProductName, item.Quantity, item.UnitPrice, item.Total, item.RentalPeriod)
		if err != nil {
			return 0, err
		}
	}

	if err := tx.Commit(); err != nil {
		return 0, err
	}

	return invoiceID, nil
}

func (r *InvoiceRepo) GetByID(id int64) (*models.Invoice, error) {
	inv := &models.Invoice{}
	var customerID sql.NullInt64
	err := r.db.QueryRow(`
		SELECT id, invoice_no, customer_id, date, subtotal, tax_rate, tax_amount, discount, total, notes, status
		FROM invoices WHERE id = ?
	`, id).Scan(&inv.ID, &inv.InvoiceNo, &customerID, &inv.Date, &inv.Subtotal, &inv.TaxRate, &inv.TaxAmount, &inv.Discount, &inv.Total, &inv.Notes, &inv.Status)
	if err != nil {
		return nil, err
	}
	if customerID.Valid {
		inv.CustomerID = &customerID.Int64
	}

	items, err := r.getItems(id)
	if err != nil {
		return nil, err
	}
	inv.Items = items

	if inv.CustomerID != nil {
		customerRepo := NewCustomerRepo(r.db)
		customer, err := customerRepo.GetByID(*inv.CustomerID)
		if err == nil {
			inv.Customer = customer
		}
	}

	return inv, nil
}

func (r *InvoiceRepo) GetAll() ([]models.Invoice, error) {
	rows, err := r.db.Query(`
		SELECT id, invoice_no, customer_id, date, subtotal, tax_rate, tax_amount, discount, total, notes, status
		FROM invoices ORDER BY created_at DESC
	`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var invoices []models.Invoice
	for rows.Next() {
		var inv models.Invoice
		var customerID sql.NullInt64
		if err := rows.Scan(&inv.ID, &inv.InvoiceNo, &customerID, &inv.Date, &inv.Subtotal, &inv.TaxRate, &inv.TaxAmount, &inv.Discount, &inv.Total, &inv.Notes, &inv.Status); err != nil {
			return nil, err
		}
		if customerID.Valid {
			inv.CustomerID = &customerID.Int64
		}
		invoices = append(invoices, inv)
	}
	return invoices, rows.Err()
}

func (r *InvoiceRepo) UpdateStatus(id int64, status models.InvoiceStatus) error {
	_, err := r.db.Exec("UPDATE invoices SET status=? WHERE id=?", status, id)
	return err
}

func (r *InvoiceRepo) Delete(id int64) error {
	_, err := r.db.Exec("DELETE FROM invoices WHERE id=?", id)
	return err
}

func (r *InvoiceRepo) GetNextInvoiceNo(prefix string) (string, error) {
	var num int
	err := r.db.QueryRow(`
		SELECT COALESCE(MAX(CAST(SUBSTR(invoice_no, ?+1) AS INTEGER)), 0) + 1 FROM invoices
	`, len(prefix)).Scan(&num)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("%s%d", prefix, num), nil
}

func (r *InvoiceRepo) getItems(invoiceID int64) ([]models.InvoiceItem, error) {
	rows, err := r.db.Query(`
		SELECT id, invoice_id, product_id, product_name, quantity, unit_price, total, rental_period
		FROM invoice_items WHERE invoice_id = ?
	`, invoiceID)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var items []models.InvoiceItem
	for rows.Next() {
		var item models.InvoiceItem
		var productID sql.NullInt64
		if err := rows.Scan(&item.ID, &item.InvoiceID, &productID, &item.ProductName, &item.Quantity, &item.UnitPrice, &item.Total, &item.RentalPeriod); err != nil {
			return nil, err
		}
		if productID.Valid {
			item.ProductID = &productID.Int64
		}
		items = append(items, item)
	}
	return items, rows.Err()
}

File: electron/database/invoice-repo.ts
import { Database as SqlJsDatabase } from 'sql.js'
import { Invoice, InvoiceItem, InvoiceStatus } from '../types'
import { all, get, run, lastInsertId } from './query'
import { SettingsRepo } from './settings-repo'

export class InvoiceRepo {
  private db: SqlJsDatabase
  private settingsRepo: SettingsRepo

  constructor(db: SqlJsDatabase) {
    this.db = db
    this.settingsRepo = new SettingsRepo(db)
  }

  getAll(): Invoice[] {
    return all<Invoice>(this.db,
      `SELECT i.*, c.name as customer_name
       FROM invoices i LEFT JOIN customers c ON i.customer_id = c.id
       ORDER BY i.created_at DESC`)
  }

  getById(id: number): Invoice | undefined {
    const invoice = get<Invoice>(this.db,
      `SELECT i.*, c.name as customer_name
       FROM invoices i LEFT JOIN customers c ON i.customer_id = c.id
       WHERE i.id = ?`, [id])
    if (!invoice) return undefined
    invoice.items = all<InvoiceItem>(this.db,
      'SELECT * FROM invoice_items WHERE invoice_id = ?', [id])
    return invoice
  }

  create(data: {
    customer_id: number | null
    date: string
    subtotal: number
    tax_rate: number
    tax_amount: number
    discount: number
    total: number
    notes: string
    items: Omit<InvoiceItem, 'id' | 'invoice_id'>[]
  }): number {
    const prefix = this.settingsRepo.get('invoice_prefix') || 'INV-'
    const nextNum = parseInt(this.settingsRepo.get('next_number') || '1', 10)
    const invoiceNo = `${prefix}${nextNum}`

    this.db.exec('BEGIN TRANSACTION')
    try {
      run(this.db,
        `INSERT INTO invoices (invoice_no, customer_id, date, subtotal, tax_rate, tax_amount, discount, total, notes, status)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 'unpaid')`,
        [invoiceNo, data.customer_id, data.date, data.subtotal, data.tax_rate,
         data.tax_amount, data.discount, data.total, data.notes])

      const invoiceId = lastInsertId(this.db)

      for (const item of data.items) {
        run(this.db,
          `INSERT INTO invoice_items (invoice_id, product_id, product_name, quantity, unit_price, total, rental_period)
           VALUES (?, ?, ?, ?, ?, ?, ?)`,
          [invoiceId, item.product_id, item.product_name, item.quantity,
           item.unit_price, item.total, item.rental_period])

        if (item.product_id && !item.rental_period) {
          run(this.db, 'UPDATE products SET stock_qty = stock_qty - ? WHERE id = ?',
            [item.quantity, item.product_id])
        }
      }

      this.settingsRepo.set('next_number', String(nextNum + 1))
      this.db.exec('COMMIT')
      return invoiceId
    } catch (e) {
      this.db.exec('ROLLBACK')
      throw e
    }
  }

  updateStatus(id: number, status: InvoiceStatus): void {
    run(this.db, 'UPDATE invoices SET status = ? WHERE id = ?', [status, id])
  }

  delete(id: number): void {
    run(this.db, 'DELETE FROM invoices WHERE id = ?', [id])
  }

  getRecent(limit: number = 5): Invoice[] {
    return all<Invoice>(this.db,
      `SELECT i.*, c.name as customer_name
       FROM invoices i LEFT JOIN customers c ON i.customer_id = c.id
       ORDER BY i.created_at DESC LIMIT ?`, [limit])
  }

  getStats(): { total: number; totalSales: number; totalPaid: number; totalUnpaid: number } {
    const result = get<{ total: number; totalSales: number; totalPaid: number; totalUnpaid: number }>(this.db,
      `SELECT
        COUNT(*) as total,
        COALESCE(SUM(total), 0) as totalSales,
        COALESCE(SUM(CASE WHEN status = 'paid' THEN total ELSE 0 END), 0) as totalPaid,
        COALESCE(SUM(CASE WHEN status = 'unpaid' THEN total ELSE 0 END), 0) as totalUnpaid
       FROM invoices`)
    return result || { total: 0, totalSales: 0, totalPaid: 0, totalUnpaid: 0 }
  }
}
